library/configure_kernel.py

```diff
@@ -0,0 +1,219 @@
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+import os
+import subprocess
+import traceback
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+class ReadKernelArgsError(Exception):
+    pass
+
+
+class UpdateKernelError(Exception):
+    pass
+
+
+class SelectCpuPartitioningError(Exception):
+    pass
+
+
+def _exec_cmd(args):
+    proc = subprocess.Popen(
+        args, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    )
+    out, err = proc.communicate()
+    return proc.returncode, out, err
+
+
+def _get_cpu_list(pci_addresses):
+    cores = []
+    for addr in pci_addresses:
+        local_cpu_list = _get_nic_cpus_without_zero_core(addr)
+        if local_cpu_list not in cores:
+            cores.append(local_cpu_list)
+    return ','.join(cores)
+
+
+def _get_nic_cpus_without_zero_core(pci_address):
+    local_cpu_list = _get_nic_cpu_list(pci_address)
+    if _is_first_core_zero(local_cpu_list):
+        local_cpu_list = _remove_first_core(local_cpu_list)
+    return local_cpu_list
+
+
+def _get_nic_cpu_list(pci_address):
+    DEVICE_PATH_FMT = '/sys/bus/pci/devices/{}'
+
+    local_cpulist = os.path.join(
+        DEVICE_PATH_FMT.format(pci_address), 'local_cpulist'
+    )
+    with open(local_cpulist) as f:
+        return f.read()
+
+
+def _is_first_core_zero(cores):
+    return cores[:1] == '0'
+
+
+def _remove_first_core(cores):
+    if cores[1] == '-':
+        return '1' + cores[1:]
+    elif cores[1] == ',':
+        return cores[2:]
+    else:
+        return ""
+
+
+def _get_default_kernel():
+    proc = subprocess.Popen(['grubby', '--default-kernel'],
+                            stdout=subprocess.PIPE)
+    return proc.stdout.read().strip()
+
+
+def _add_hugepages(kernel):
+    if _current_hugepages():
+        return
+
+    args = 'default_hugepagesz=2M hugepagesz=2M hugepages=512'
+    proc = subprocess.Popen(['grubby', '--args="{}"'.format(args),
+                             '--update-kernel', kernel])
+    out, err = proc.communicate()
+    if err:
+        raise UpdateKernelError(out)
+    return True
+
+
+def _add_isolated_cpus(cpu_list):
+    VARIABLES_FILE = '/etc/tuned/cpu-partitioning-variables.conf'
+
+    changed = False
+    new_lines = []
+    with open(VARIABLES_FILE) as f:
+        lines = f.readlines()
+        for line in lines:
+            if line.startswith('isolated_cores'):
+                required_line = 'isolated_cores={}'.format(cpu_list)
+                if line != required_line:
+                    line = required_line
+                    changed = True
+            new_lines.append(line)
+
+    with open(VARIABLES_FILE, 'w') as f:
```

Comment:
does Ansible has means for a safe write? this can fail and eliminate your VARIABLE_FILES.

Reply:
I can't find anything. I could use a temp file in the same filesystem and copy it over it over, but is the risk worth the added code complexity?